common/aws/s3/fragment.go

```diff
@@ -0,0 +1,128 @@
+package s3
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+// GetFragmentCount returns the number of fragments that a file of the given size will be broken into.
+func GetFragmentCount(fileSize int, fragmentSize int) int {
+	if fileSize < fragmentSize {
+		return 1
+	} else if fileSize%fragmentSize == 0 {
+		return fileSize / fragmentSize
+	} else {
+		return fileSize/fragmentSize + 1
+	}
+}
+
+// GetFragmentKey returns the key for the fragment at the given index.
+//
+// Fragment keys take the form of "prefix/body-index[f]". The prefix is the first prefixLength characters
+// of the file key. The body is the file key. The index is the index of the fragment. The character "f" is appended
+// to the key of the last fragment in the series.
+//
+// Example: fileKey="abc123", prefixLength=2, fragmentCount=3
+// The keys will be "ab/abc123-0", "ab/abc123-1", "ab/abc123-2f"
```

Comment:
why is the prefix needed?

Reply:
S3 throttles bandwidth based on prefix. That means reading and writing keys with two different prefixes has an aggregate upper bandwidth limit twice as high as reading and writing keys with a single prefix.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html

> For example, your application can achieve at least 3,500 PUT/COPY/POST/DELETE or 5,500 GET/HEAD requests per second per partitioned Amazon S3 prefix. There are no limits to the number of prefixes in a bucket. You can increase your read or write performance by using parallelization. For example, if you create 10 prefixes in an Amazon S3 bucket to parallelize reads, you could scale your read performance to 55,000 read requests per second. Similarly, you can scale write operations by writing to multiple prefixes.

The goal here is to spread the data into different prefixes so that we have a much higher upper limit on bandwidth.

Reply:
This is cool, didn't know about it. But I'm still a bit puzzled by how you are making use of this though. Given that the prefix length is a global config that is always the same, then all fragments of a file will have the same prefix right? Compare
```
"ab/abc123-0", "ab/abc123-1", "ab/abc123-2f"
```
to
```
"abc123-0", "abc123-1", "abc123-2f"
```

All fragments in both cases have the same prefix, so I don't quite understand how using an extra prefix is helping here. Am I missing something?

Reply:
+1 
The prefix helps with creating virtual hierarchy within the bucket, but all files with the same prefix is already in the same bucket, right?

Reply:
All files end up in the same bucket. The prefix has nothing to do with which bucket each file is written to. Our use of prefixes is entirely a tactic to improve our bandwidth with S3, not as a mechanism used to organize our data for our own purposes.

Reply:
I see. Ok so the idea is that all files are in the same bucket, which means without prefixes all reads/writes to this bucket would have a total bandwidth limit. Since we want a per-file bandwidth limit, you add these prefixes.

Why don't you just use "-" as the prefix delimiter then, instead of adding this prefix? Still feels to me like the real useful prefix is the file name, not this deterministically shortened version of it (which I still think isn't useful), no?

On a side note, how exactly are buckets organized? When do we create a new bucket, if ever?